### PR TITLE
Add Timestamp(Ms,Us)Attribute

### PR DESCRIPTION
**Pull request by ikonst**

### CODE REVIEW

**Comment by ikonst** [comment]
(need to merge #9 to pass mypy)

File: pynamodb_attributes/timestamp.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        return datetime.fromtimestamp(int(value) / self._multiplier, tz=timezone.utc)
+
+    def serialize(self, value: datetime) -> str:
+        return str(int(value.timestamp() * self._multiplier))
+
+    def __set__(self, instance, value: datetime) -> None:
+        if not isinstance(value, datetime):
+            raise TypeError(f"value has invalid type '{type(value)}; datetime expected")
+        if value.tzinfo is None:
+            raise TypeError("offset-aware datetime expected")
```
Comment (nolanlum) [inline pynamodb_attributes/timestamp.py]: I think most documentation uses "aware" or "timezone aware" to describe this situation

File: pynamodb_attributes/timestamp.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def deserialize(self, value: str) -> datetime:
+        return datetime.fromtimestamp(int(value) / self._multiplier, tz=timezone.utc)
+
+    def serialize(self, value: datetime) -> str:
+        return str(int(value.timestamp() * self._multiplier))
+
+    def __set__(self, instance, value: datetime) -> None:
+        if not isinstance(value, datetime):
+            raise TypeError(f"value has invalid type '{type(value)}; datetime expected")
+        if value.tzinfo is None:
```
Comment (nolanlum) [inline pynamodb_attributes/timestamp.py]: turns out that [not only does `tzinfo` need to be non-`None`, but `tzinfo.utcoffset` must also return non-`None`](https://stackoverflow.com/questions/5802108/how-to-check-if-a-datetime-object-is-localized-with-pytz).

File: pynamodb_attributes/timestamp.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    def deserialize(self, value: str) -> datetime:
+        return datetime.fromtimestamp(int(value) / self._multiplier, tz=timezone.utc)
+
+    def serialize(self, value: datetime) -> str:
+        return str(int(value.timestamp() * self._multiplier))
+
+    def __set__(self, instance, value: datetime) -> None:
+        if not isinstance(value, datetime):
+            raise TypeError(f"value has invalid type '{type(value)}; datetime expected")
```
Comment (nolanlum) [inline pynamodb_attributes/timestamp.py]: missing a `'` in the exception message here

File: pynamodb_attributes/timestamp.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        if value.tzinfo is None:
+            raise TypeError("offset-aware datetime expected")
+        return super().__set__(instance, value)
+
+
+class TimestampMsAttribute(TimestampAttribute):
+    """"
+    Stores times as Unix epoch timestamps in milliseconds in a DynamoDB number.
+    """
+    _multiplier = 1_000.0
```
Comment (nolanlum) [inline pynamodb_attributes/timestamp.py]: underscores in number literals were only added in python 3.6, I'd leave them out for compatibility.

File: pynamodb_attributes/timestamp.py (hunk 1 above)
Comment (ikonst) [reply pynamodb_attributes/timestamp.py]: Alright, will change to "aware".

I was going off:
```
>>> from datetime import datetime, timezone
>>> datetime.now(tz=timezone.utc) < datetime.utcnow()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't compare offset-naive and offset-aware datetimes
```

File: pynamodb_attributes/timestamp.py (hunk 2 above)
Comment (ikonst) [reply pynamodb_attributes/timestamp.py]: 😮

File: pynamodb_attributes/timestamp.py (hunk 1 above)
Comment (nolanlum) [reply pynamodb_attributes/timestamp.py]: amazingly inconsistent I love it

**Comment by ikonst** [comment]
Since `datetime` doesn't have nanosecond precision, maybe I should have a separate attribute that converts nanoseconds to a `google.protobuf.internal.well_known_types.Timestamp` or something (with a loose dependency on Protobuf).

**Review by nolanlum — APPROVED** [review]

**Pull request merged**